services/venice-controller/src/main/java/com/linkedin/venice/controller/kafka/protocol/serializer/NewSemanticUsageValidator.java

```diff
@@ -0,0 +1,253 @@
+package com.linkedin.venice.controller.kafka.protocol.serializer;
+
+import com.linkedin.avroutil1.compatibility.AvroCompatibilityHelper;
+import com.linkedin.venice.utils.AvroSchemaUtils;
+import com.linkedin.venice.utils.Pair;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.BiFunction;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericRecord;
+
+
+/**
+ * Provides a semantic validation function to check whether an object adheres to a given schema.
+ * An object is considered valid if:
+ * <ol>
+ *   <li>It is null.</li>
+ *   <li>It is a nested type and equals its schema's default value.</li>
+ *   <li>It is a non-nested field and matches either the schema's default value or its type's default.</li>
+ * </ol>
+ *
+ * If the object is not using new semantic, we return False. Otherwise, we return True along with the message
+ * in errorMessage. The traverser should stop traversing the object if the return value is True.
+ */
+public class NewSemanticUsageValidator {
```

Comment:
Thanks for working on this functionality! IMHO, this is a fairly generic utility and should not be Venice-specific.

It is fine for it to start as such, but once it is stabilized and we feel ready, it may be more appropriate to contribute it to [avro-util](https://github.com/linkedin/avro-util), or maybe even Apache Avro itself.

Reply:
Thank you. When the class is stable, I will consider to bring it to more general place.